text/0011-html_viz.md

```diff
@@ -0,0 +1,177 @@
+
+### Vizualizations in datasets
+
+Saving a dataset will by default execute the default template to a file called `index.html` & place it in the IPFS merkle-DAG. When an IPFS HTTP gateway receives a request for a DAG that is a directory containing `index.html`, it returns the HTML file by default. This means when a user visits a dataset on the d.web–completely outside the Qri system of dataset production–they are greeted with a well-formatted dataset document by default.
+
+While care will be taken to keep `index.html` files small, users may understandably want to disable them entirely. To achieve this we'll add a new flag to `qri save` and `qri update`: `--no-render`. No render will prevent the execution of any viz template. This will save a few KB from version to version at the cost of usability.
+
+Users can _override_ the default template by supplying their own custom viz templates either by specifying a `viz.scriptPath`:
+
+`dataset.yaml`:
+```yaml
+name: example_dataset
+# additional fields elided ...
+viz:
+  syntax: html
+  scriptPath: template.html
+```
+
+and running save:
+```
+$ qri save --file dataset.yaml
+```
+
+Or by running `qri save` with an `.html` file argument:
+`qri save --file template.html me/example_dataset`
+
+Since the above example provided no additional configuration details for the `viz` component in `dataset.yaml`, the two calls will have the same effect.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Template API
+
+Introducing HTML template execution requires defining an API for template values. This API will need to be documented & maintained just like any other API in the Qri ecosystem.
+
+The template implementation will use the [go language html template package](https://golang.org/pkg/html/template)
+
+#### Dataset is exposed as `ds`
+
+HTML template should expose a dataset document at `ds`. By exposing the document as `ds`, it matches our for referring to a dataset in starlark., and allows access to all defined parts of a dataset. ds should use _lowercase_ fields for component accessors. eg:
+
+```html
+{{ ds.meta.title }}
+```
+
+Undefined components should be defined as empty struct pointers if null. For example a dataset _without_ a transform the following template should print nothing:
+```html
+{{ ds.transform }}
+```
+And this should error:
+```html
+{{ ds.transform.script }}
+```
+
+Having default empty pointers prevents unnecessary `if` clauses, allowing a skip to tests for existence on component fields:
+```html
+{{ if ds.transform.script }}
+{{ end }}
+```
+
+### Template functions
+
+Top level functions should be loaded into the template `funcs` space to make rendering templates easier. The go html template package comes with [predefined functions](https://golang.org/pkg/text/template/#hdr-Functions), all of which are included in this RFC. An example prints the length of the body:
```

Comment:
I don't totally understand what you mean by "all of which are included in this RFC"

Do you mean that there are examples of all of those functions someone in this RFC, or that this RFC proposes adding them all into our templating?

Reply:
Good point, I meant the latter. That this RFC will de-facto introduce all of these functions into our template engine unless we explicitly turn them off (because we're using the template package), and I don't think we should turn them off.